### ISSUE TITLE
"Error reading configuration" on multiple actions

### ISSUE THREAD

**Issue by renard** [seveas/python-hpilo#108]
I run `python-hpilo` (the git master version) on ILO4 (version 2.40) to reset the device to factory default and put a license key using following command (Note I have an official working key):

```
hpilo_cli -d -d  -d -plocal localhost factory_defaults + activate_license key=123456789ABCDEFGHIJKLMNOP
```

This fails with following output:

```
Launching hponcfg
Sending XML request, 315 bytes
<?xml version="1.0"?>
<RIBCL
 VERSION="2.0"
>
<LOGIN
 PASSWORD="********"
 USER_LOGIN="Administrator"
>
<RIB_INFO
 MODE="write"
>
<FACTORY_DEFAULTS
 />
</RIB_INFO>
<RIB_INFO
 MODE="write"
>
<LICENSE
>
<ACTIVATE
 KEY="123456789ABCDEFGHIJKLMNOP"
 />
</LICENSE>
</RIB_INFO>
</LOGIN>
</RIBCL>

Received 1438 bytes


<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
<INFORM>Integrated Lights-Out will reset at the end of the script.</INFORM>
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x008F"
    MESSAGE='Error reading configuration'
     />
</RIBCL>Traceback (most recent call last):
  File "/usr/bin/hpilo_cli", line 326, in <module>
    main()
  File "/usr/bin/hpilo_cli", line 210, in main
    results = ilo.call_delayed()
  File "/usr/lib/python2.7/dist-packages/hpilo.py", line 768, in call_delayed
    header, message = self._request(root)
  File "/usr/lib/python2.7/dist-packages/hpilo.py", line 252, in _request
    message = self._parse_message(data)
  File "/usr/lib/python2.7/dist-packages/hpilo.py", line 588, in _parse_message
    raise IloError(message, status)
hpilo.IloError: Error reading configuration
```

On the other hand, if I run the commands separately this works like a charm. If I put the generated XML in a file and run `hponfcg -f file.xml` this also fails:

```
HP Lights-Out Online Configuration utility
Version 4.4.0 Date 06/13/2014 (c) Hewlett-Packard Company, 2014
Firmware Revision = 2.40 Device type = iLO 4 Driver name = hpilo
<INFORM>Integrated Lights-Out will reset at the end of the script.</INFORM>
</-- ERROR :      STATUS= 0x008F
     MESSAGE= Error reading configuration -->

Please wait while the firmware is reset. This might take a minute / ^[
Script failed
```

Buggy XML:
```xml
<?xml version="1.0"?>
<RIBCL VERSION="2.0">
<LOGIN PASSWORD="********" USER_LOGIN="Administrator">
  <RIB_INFO MODE="write">
    <FACTORY_DEFAULTS />
   </RIB_INFO>
  <RIB_INFO MODE="write">
    <LICENSE>
      <ACTIVATE KEY="123456789ABCDEFGHIJKLMNOP" />
    </LICENSE>
  </RIB_INFO>
</LOGIN>
</RIBCL>
```
If I put both `FACTORY_DEFAULTS` and `LICENSE` command in the same `RIB_INFO` the XML is working:

Working xml:
```xml
<?xml version="1.0"?>
<RIBCL VERSION="2.0">
<LOGIN PASSWORD="********" USER_LOGIN="Administrator">
  <RIB_INFO MODE="write">
    <FACTORY_DEFAULTS />
    <LICENSE>
      <ACTIVATE KEY="123456789ABCDEFGHIJKLMNOP" />
    </LICENSE>
  </RIB_INFO>
</LOGIN>
</RIBCL>
```

output:

```
HP Lights-Out Online Configuration utility
Version 4.4.0 Date 06/13/2014 (c) Hewlett-Packard Company, 2014
Firmware Revision = 2.40 Device type = iLO 4 Driver name = hpilo
<INFORM>Integrated Lights-Out will reset at the end of the script.</INFORM>

Please wait while the firmware is reset. This might take a minute
Script succeeded
```

Did I miss something?

Thanks for your support.

Cheers

**Comment by renard** [comment]
This is amazing because if I run `factory_defaults` and  `set_server_auto_pwr` in a row, this works:

```
hpilo_cli -d -d  -d -plocal localhost factory_defaults + set_server_auto_pwr setting=Random
Launching hponcfg
Sending XML request, 286 bytes
<?xml version="1.0"?>
<RIBCL
 VERSION="2.0"
>
<LOGIN
 PASSWORD="********"
 USER_LOGIN="Administrator"
>
<RIB_INFO
 MODE="write"
>
<FACTORY_DEFAULTS
 />
</RIB_INFO>
<SERVER_INFO
 MODE="write"
>
<SERVER_AUTO_PWR
 VALUE="Random"
 />
</SERVER_INFO>
</LOGIN>
</RIBCL>

Received 1770 bytes


<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
<INFORM>Integrated Lights-Out will reset at the end of the script.</INFORM>
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
</RIBCL>>>> my_ilo.factory_defaults()
>>> my_ilo.set_server_auto_pwr(setting="Random")
```

**Comment by renard** [comment]
FYI this is a piece of XML I run for several years to configure my ILO using `hponcfg`:

```xml
<RIBCL VERSION="2.0">

  <LOGIN USER_LOGIN="admin" PASSWORD="password">

    <RIB_INFO MODE="write">
      <FACTORY_DEFAULTS/>
    </RIB_INFO>

    <RIB_INFO MODE="write">
      <LICENSE>
	<ACTIVATE KEY="@@ILO_LICENSE@@"/>
      </LICENSE>
    </RIB_INFO>

    <SERVER_INFO MODE="write">
      <SERVER_AUTO_PWR VALUE="Random"/>
    </SERVER_INFO>

    <USER_INFO mode="write">
      <ADD_USER USER_NAME="Nagios Monitor"
    		USER_LOGIN="@@ILO_NAGIOS_USER@@"
    		PASSWORD="@@ILO_NAGIOS_PASSWORD@@">
    	<ADMIN_PRIV value="N"/>
    	<REMOTE_CONS_PRIV value="N"/>
    	<RESET_SERVER_PRIV value="N"/>
    	<VIRTUAL_MEDIA_PRIV value="N"/>
    	<CONFIG_ILO_PRIV value="N"/>
      </ADD_USER>

      <ADD_USER USER_NAME="root"
    		USER_LOGIN="@@ILO_ROOT_USER@@"
    		PASSWORD="@@ILO_ROOT_PASSWORD@@">
    	<ADMIN_PRIV value="Y"/>
    	<REMOTE_CONS_PRIV value="Y"/>
    	<RESET_SERVER_PRIV value="Y"/>
    	  <VIRTUAL_MEDIA_PRIV value="Y"/>
    	  <CONFIG_ILO_PRIV value="Y"/>
      </ADD_USER>
    </USER_INFO>

    <RIB_INFO mode="write">
      <MOD_NETWORK_SETTINGS>
    	<ENABLE_NIC VALUE="Y"/>
    	<SPEED_AUTOSELECT VALUE="Y"/>
    	<NIC_SPEED VALUE="10"/>
    	<FULL_DUPLEX VALUE="N"/>
    	<DHCP_ENABLE VALUE="N"/>
    	<DHCP_GATEWAY VALUE="N"/>
    	<DHCP_DNS_SERVER VALUE="N"/>
    	<DHCP_WINS_SERVER VALUE="N"/>
    	<DHCP_STATIC_ROUTE VALUE="N"/>
    	<DHCP_DOMAIN_NAME VALUE="N"/>
    	<REG_WINS_SERVER VALUE="Y"/>
    	<REG_DDNS_SERVER VALUE="Y"/>
    	<PING_GATEWAY VALUE="Y"/>
    	<IP_ADDRESS VALUE="@@ILO_IP_ADDRESS@@"/>
    	<SUBNET_MASK VALUE="@@ILO_IP_NETMASK@@"/>
    	<GATEWAY_IP_ADDRESS VALUE="@@ILO_IP_GATEWAY@@"/>
    	<DNS_NAME VALUE="@@ILO_HOSTNAME@@"/>
    	<DOMAIN_NAME VALUE="@@ILO_DOMAIN@@"/>
    	<PRIM_DNS_SERVER VALUE="@@ILO_DNS_SERVER1@@"/>
    	<SEC_DNS_SERVER VALUE="@@ILO_DNS_SERVER2@@"/>
    	<TER_DNS_SERVER VALUE="0.0.0.0"/>
    	<PRIM_WINS_SERVER VALUE="0.0.0.0"/>
    	<SEC_WINS_SERVER VALUE="0.0.0.0"/>
	<STATIC_ROUTE_1 DEST = "0.0.0.0" GATEWAY = "0.0.0.0"/>
	<STATIC_ROUTE_2 DEST = "0.0.0.0" GATEWAY = "0.0.0.0"/>
	<STATIC_ROUTE_3 DEST = "0.0.0.0" GATEWAY = "0.0.0.0"/>
      </MOD_NETWORK_SETTINGS>
    </RIB_INFO>

  </LOGIN>
</RIBCL>
```

And as you can notice, the first commands are very similar as the XML you generated `FACTORY_DEFAULTS` and `LICENSE` in their own `RIB_INFO` and this works.

I know HP XML parser is a bit capricious maybe I hit a ILO bug or something.

Cheers.

**Comment by seveas** [comment]
Might be an iLO bug, it might be documented that you can have at most one RIB_INFO block when writing, or especially when resetting.

I think it makes sense to merge these blocks if possible. And maybe warn if the generated xml has 2 rib_info blocks separated by a server_info block and can thus not be merged.

Does your last XML sample work on servers where you hit this bug? It also has separate RIB_INFO blocks like the XML generated by python-hpilo, so I guess it doesn't.

**Comment by renard** [comment]
Yes that's is what lead me to an ILO bug. I use my XML sample on ILO 2 to 4 without any issue.

It seems that on Gen9 servers, my XML does not really work either. for it to work I have to merge all `RIB_INFO`.

Still don't know why.

**Comment by renard** [comment]
BTW I notice the problem with newer ilo firmware versions

**Comment by renard** [comment]
@seveas I can confirm that in ILO 4 v2.31 everything works fine. If you upgrade to ilo v2.40 it fails.

TIP: If you need to downgrade an ilo version use the `--force` option to the scexe.

**Comment by seveas** [comment]
Ok, I'm going to close this bug then, as this is not a python-hpilo bug. Please contact HP about this.

**Comment by renard** [comment]
Ok thank you.
Maybe you can just add a note about that in the README.

**Comment by seveas** [comment]
Good idea, added in fbbc7c3

**Comment by kryptohash** [comment]
@renard 
I have some experience dealing with iLO scripting.  Basically, you are executing two commands in one XML script that you need to put in two separate XML scripts:

The <FACTORY_DEFAULTS /> tag causes iLO to erase its NVRAM and schedule a reboot.  Now, here is the catch:  The iLO reboot starts as soon as the XML parser sees the closing </RIB_INFO> tag.   Anything you add after that </RIB_INFO> tag may not execute properly because, even though the XML parser keeps running your script, iLO behind the scenes is already preparing itself to go for a reboot and sometimes iLO reboots really quick.    

Because of this behavior, I've seen cases where some iLO settings could get corrupted.  I had this problem few years ago and I contacted HP support.  They told me to separate the XML scripts,or change the order of Tags in order to leave the one that triggers the iLO reboot to the end.

These scripts below will always work.  But, you need to wait until iLO comes back from reboot which could take ~1 minute.

<?xml version="1.0"?>
<RIBCL VERSION="2.0">
 <LOGIN PASSWORD="********" USER_LOGIN="Administrator">
  <RIB_INFO MODE="write">
    <FACTORY_DEFAULTS />
   </RIB_INFO>
 </LOGIN>
</RIBCL>

<?xml version="1.0"?>
<RIBCL VERSION="2.0">
 <LOGIN PASSWORD="********" USER_LOGIN="Administrator">
  <RIB_INFO MODE="write">
    <LICENSE>
      <ACTIVATE KEY="123456789ABCDEFGHIJKLMNOP" />
    </LICENSE>
  </RIB_INFO>
 </LOGIN>
</RIBCL>

**Comment by renard** [comment]
@kryptohash again my script works nicely with up to version 2.31 and fails with version 2.40. I use it doe years.
I noticed it fails since I upgraded to 2.40.

**Comment by kryptohash** [comment]
@renard 
I know your script used to work but, it doesn't anymore.  I had the same problem few years ago when a couple of my iLOs would end up with corrupted SNMP settings.  
My script had multiple RIB_INFO sections just like your script and, since the first section of my script was making changes to iLO network settings (MOD_NETWORK_SETTINGS), all my iLOs were  rebooting at the end of the first RIB_INFO section.  
The result was a totally unpredictable behavior where some of my iLOs were getting correct SNMP settings while others were not. My iLOs were all at the same firmware level but, on different server models. Yet, they behaved differently. It was a disaster.

HP support told me to never add extra RIBCL sections after one that is known to trigger an iLO reboot. Any MOD_NETWORK_SETTINGS change will cause iLO to reboot so, will FACTORY_DEFAULTS by the way.  Make the change to my script solved my problems for good.

Basically, HP support told me that if you see this response coming from iLO:

```
<?xml version="1.0"?>
<RIBCL VERSION="2.23">
<RESPONSE
    STATUS="0x0000"
    MESSAGE='No error'
     />
<INFORM>Integrated Lights-Out will reset at the end of the script.</INFORM>
</RIBCL>
```

Any RIBCL section you see executing after this message is not guaranteed to succeed because, iLO is in the process of shutting down. In other words, iLO has began closing its doors and windows and, turning off the lights one by one in order to go for a reboot.

**Comment by seveas** [comment]
In commit 2469fed, on the tag-merge branch, I've implemented tag merging. This should make factory_defaults and activate_license in the same XML script work again. Please test.

**Comment by renard** [comment]
@seveas This works!
BUT,
I have to setup the correct order: all rib_info should be given in a row:

- factory_defaults
- activate_license
- mod_network_settings
- add_user
- add_user
- set_server_auto_pwr

If you can catch them up and setup the correct order or this would be greater. Otherwise just let user know.

**Comment by renard** [comment]
@kryptohash Yep I noticed that HP xml parse is rotten to the core. What I don't understand is the fact they managed to make it work (for ilo 2, 3 and 4) and now they broke it.

**Comment by seveas** [comment]
@renard reordering calls is a no-go, and I'm not even sure about the warning (except a warning in the docs).

**Comment by renard** [comment]
@seveas Ok as you want. Anyway it works now, thanks for the support.

**Comment by seveas** [comment]
Added in 786283c
